stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+  guard let tracedFn = _RuntimeConfig.traceState.getFunction else {
+    fatalError("Not in tracing mode!.")
+  }
+  _RuntimeConfig.traceState = .notTracing
+
+  return tracedFn.finalizeAndExecute(traceeBasicName: fnName,
+                                     inputs: inputs,
+                                     outputs: outputs)
+}
+
+public extension TensorArrayProtocol {
+  // The returned handles are owned by the caller.
+  func _getCTensorHandles() -> [CTensorHandle] {
+    debugLog("Getting \(self._tensorHandleCount) C handles.")
+    let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(
+      capacity: Int(self._tensorHandleCount))
+    debugLog("Unpacking handles into buffer.")
+    self._unpackTensorHandles(into: buffer)
+    let status = TF_NewStatus()
+    debugLog("Copying buffer content to output handles.")
+    var output: [CTensorHandle] = []
+    for idx in 0..<self._tensorHandleCount {
+      let address = buffer.advanced(by: Int(idx))
+      let isConcrete = TFE_TensorHandleIsConcrete(address.pointee) != 0
+      debugLog("""
+                 Copying the \(idx)-th C handle \(address.pointee) with \
+                 concrete=\(isConcrete).
+                 """)
+      let newHandle = TFE_TensorHandleCopySharingTensor(address.pointee,
+                                                        status)
+      checkOk(status)
+      debugLog("""
+                 Copying the \(idx)-th C handle \(address.pointee) with \
+                 concrete=\(isConcrete) to C handle create \(newHandle!).
+                 """)
+      output.append(newHandle!)
+    }
+    TF_DeleteStatus(status)
+    return output
+  }
+}
+
+public extension TensorGroup {
+  // The tensors in `input` are NOT owned by this instance.
+  init(_copying input: [CTensorHandle]) {
+    assert(Self._tensorHandleCount == input.count)
+    let buffer = UnsafeMutablePointer<CTensorHandle>.allocate(capacity: input.count)
+    let status = TF_NewStatus()
+    // copy input to buffer
+    for (idx, inputTensorHandle) in input.enumerated() {
+      let address = buffer.advanced(by: idx)
+      // Each tensor can be symbolic (e.g. when using this API to create a
+      // symbolic input instance to tracee) or concrete (e.g. when creating the
+      // final output of the tracee).
+      let newHandle = TFE_TensorHandleCopySharingTensor(inputTensorHandle, status)
+      checkOk(status)
+      address.initialize(to: newHandle!)
+    }
+    TF_DeleteStatus(status)
+    self.init(_owning: buffer)
+  }
+}
+
+// TODO: assess if this protocol should be folded into TensorArrayProtocol.
+public protocol TensorArrayProtocolEnhanced : TensorArrayProtocol {
+  // Create an instance based on `inputs`, which can be symbolic (e.g. when
+  // creating a symbolic input to tracee) or concrete (e.g. when creating a
+  // final output of executing the tracee).
+  func createInstance(_owning inputs: [CTensorHandle]) -> Self
```

Comment:
* It's fine to add this to `TensorArrayProtocol` today (and remove `TensorArrayProtocolEnhanced`). Otherwise generic algorithms won't be compatible with the `trace(with:in:)` operator. `TensorArrayProtocol` is the public API that users interact with.
* Does this rely on `self`, i.e. do `inputs` have to match the shapes of each tensor in `self`?
  * If so, it's called a factory method in Swift, which could be called `_makeInstance(owning:)` according to the naming convention.
  * It it does not rely on `self`, this should be a static. Since the semantics of this method is initialization, it should an initializer requirement on `TensorArrayProtocol`, i.e. `init(_owning:)`.

Depending on whether it becomes an initializer or a factory method. In the initializer case, the first argument label should be underscored. In the factory method case, the method name should be underscored but the argument label need not be.

Reply:
Thanks for the style guidance. This is a factory method, and I named it accordingly.

Merging the protocols SGTM. I'm deferring that work (left a TODO), because it requires we implement this new method `_makeInstance` for each struct/class that conforms to `TensorArrayProtocol` (e.g. `Tensor._makeInstance()`).